functions/go/Makefile

```diff
@@ -1,34 +1,116 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+SHELL=/bin/bash
 GOPATH := $(shell go env GOPATH)
+TAG := latest
 
 # Edit this list to contain all go functions
 FUNCTIONS := \
 	set-namespace
 
+# Targets for running all function tests
+FUNCTION_TESTS := $(patsubst %,%-TEST,$(FUNCTIONS))
+# Targets to build functions
+FUNCTION_BUILDS := $(patsubst %,%-BUILD,$(FUNCTIONS))
+# Targets to build functions with dev tag
+FUNCTION_DEV_BUILDS := $(patsubst %,%-BUILD-DEV,$(FUNCTIONS))
+# Targets to push images
+FUNCTION_PUSH := $(patsubst %,%-PUSH,$(FUNCTIONS))
+# Current function name used by individual function targets
 CURRENT_FUNCTION := unknown
 
-.PHONY: fix vet fmt test lint all
-.PHONY: $(FUNCTIONS)
+.DEFAULT_GOAL := help
+.PHONY: help
+help: ## Print this help
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
 
-all: $(FUNCTIONS)
+.PHONY: verify
+verify: $(FUNCTIONS) ## Verify all functions
 
+.PHONY: $(FUNCTIONS)
 $(FUNCTIONS):
-	$(MAKE) CURRENT_FUNCTION=$@ function
+	$(MAKE) CURRENT_FUNCTION=$@ func-verify
+
+.PHONY: test
+test: $(FUNCTION_TESTS) ## Run unit tests for all functions
+
+.PHONY: $(FUNCTION_TESTS)
+$(FUNCTION_TESTS):
+	$(MAKE) CURRENT_FUNCTION=$(subst -TEST,,$@) func-test
+
+.PHONY: build
+build: $(FUNCTION_BUILDS) ## Build all function images. 'TAG' env is used to specify tag. 'latest' will be used if not set.
+
+.PHONY: $(FUNCTION_BUILDS)
+$(FUNCTION_BUILDS):
+	$(MAKE) CURRENT_FUNCTION=$(subst -BUILD,,$@) func-build
 
-function: fix vet fmt test lint
+.PHONY: build-dev
+build-dev: $(FUNCTION_DEV_BUILDS) ## Build all function images with tag 'dev'. This is used for local tests.
 
-fix:
+.PHONY: $(FUNCTION_DEV_BUILDS)
+$(FUNCTION_DEV_BUILDS):
+	$(MAKE) CURRENT_FUNCTION=$(subst -BUILD-DEV,,$@) TAG=dev func-build
+
+.PHONY: push
+push: $(FUNCTION_PUSH) ## Push images to registry. WARN: This operation should only be done in CI environment.
+
+.PHONY: $(FUNCTION_PUSH)
+$(FUNCTION_PUSH):
+	$(MAKE) CURRENT_FUNCTION=$(subst -PUSH,,$@) func-push
+
+# Recipes for individual function
+.PHONY: func-fix func-vet func-fmt func-test func-lint \
+	func-build func-push func-verify
+func-verify: func-fix func-vet func-fmt func-test func-lint
+
+func-fix:
 	cd $(CURRENT_FUNCTION) && go fix ./...
 
-fmt:
+func-fmt:
 	cd $(CURRENT_FUNCTION) && go fmt ./...
 
-lint:
+func-lint:
 	(which $(GOPATH)/bin/golangci-lint || \
 		curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v1.32.2)
 	cd $(CURRENT_FUNCTION) && $(GOPATH)/bin/golangci-lint run ./...
 
-test:
+func-test:
 	cd $(CURRENT_FUNCTION) && go test -cover ./...
 
-vet:
+func-vet:
 	cd $(CURRENT_FUNCTION) && go vet ./...
+
+func-build: func-verify
+	@echo Building image for $(CURRENT_FUNCTION)...
+	@( \
```

Comment:
as a general rule, it's fine to have helper bash scripts that are called by the Makefile. In general, any non-trivial scripting is very awkward when inlined in Makefile. This comes from experience working with some nightmarish Makefiles ;)